letshelp-letsencrypt/letshelp_letsencrypt/letshelp_letsencrypt_apache.py

```diff
@@ -0,0 +1,303 @@
+        modules_fd.write(proc.communicate()[0])
+
+    proc = subprocess.Popen([args.apache_ctl, "-d", args.server_root, "-f",
+                             args.config_file, "-t", "-D", "DUMP_VHOSTS"],
+                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    with open(os.path.join(tempdir, "vhosts"), "w") as vhosts_fd:
+        vhosts_fd.write(proc.communicate()[0])
+
+    return tempdir
+
+
+def verify_config(args):
+    """Verifies server_root and config_file specify a valid config
+
+    :param argparse.Namespace args: Parsed command line arguments
+
+    """
+    try:
+        with open(os.devnull, "w") as devnull:
+            subprocess.check_call([args.apache_ctl, "-d", args.server_root,
+                                   "-f", args.config_file, "-t"],
+                                  stdout=devnull, stderr=subprocess.STDOUT)
+    except OSError:
+        sys.exit(_NO_APACHECTL)
+    except subprocess.CalledProcessError:
+        sys.exit("Syntax check from apachectl failed")
+
+
+def locate_config(apache_ctl):
+    """Uses the apachectl binary to find configuration files
+
+    :param str apache_ctl: Path to `apachectl` binary
+
+
+    :returns: Path to Apache server root and main configuration file
+    :rtype: `tuple` of `str`
+
+    """
+    try:
+        proc = subprocess.Popen([apache_ctl, "-V"],
+                                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, _ = proc.communicate()
+    except OSError:
+        sys.exit(_NO_APACHECTL)
+
+    server_root = config_file = ""
+    for line in output.splitlines():
+        # Relevant output lines are of the form: -D DIRECTIVE="VALUE"
+        if "HTTPD_ROOT" in line:
+            server_root = line[line.find("\"")+1:-1]
+        elif "SERVER_CONFIG_FILE" in line:
+            config_file = line[line.find("\"")+1:-1]
+
+    if not (server_root and config_file):
+        sys.exit("Unable to locate Apache configuration. Please run this "
+                 "script again and specify --server-root and --config-file")
+
+    return server_root, config_file
+
+
+def get_args():
+    """Parses command line arguments
+
+    :returns: Parsed command line options
+    :rtype: argparse.Namespace
+
+    """
+    parser = argparse.ArgumentParser(description=_DESCRIPTION)
+    parser.add_argument("-c", "--apache-ctl", default="apachectl",
+                        help="path to the `apachectl` binary")
+    parser.add_argument("-d", "--server-root",
+                        help=("location of the root directory of your Apache "
+                              "configuration"))
+    parser.add_argument("-f", "--config-file",
+                        help=("location of your main Apache configuration "
+                              "file relative to the server root"))
+    args = parser.parse_args()
+
+    # args.server_root XOR args.config_file
+    if bool(args.server_root) != bool(args.config_file):
```

Comment:
The script needs to know the location of both the Apache server root and main configuration file. The idea was that if you wanted to submit an Apache configuration in a location other than Apache's compiled in defaults, you need to specify the location of both. I suppose it's possible you only need to include one (the location of the configuration file relative to the server root is the same) but normally you want to specify both or neither of the two arguments (and use the compiled in defaults).